src/Moq/InvocationShape.cs

```diff
@@ -91,6 +91,11 @@ private bool IsOverride(MethodInfo invocationMethod)
 				return false;
 			}
 
+			if (!invocationMethod.DeclaringType.IsAssignableFrom(method.DeclaringType))
+			{
+				return false;
+			}
```

Comment:
Unfortunately, this is not a workable solution.

Notice how, after this addition, we now have two subsequent checks: 

* `!method.DeclaringType.IsAssignableFrom(invocationMethod.DeclaringType)`
* `!invocationMethod.DeclaringType.IsAssignableFrom(method.DeclaringType)`

`IsAssignableFrom` checks for a subtype-or-same-type relationship, so these checks combined are equivalent to:

* `method.DeclaringType != invocationMethod.DeclaringType`

And this obviously is not the correct thing to do when checking for method overrides, since an overriding method *can* be declared in a type other than that of the overridden method (specifically, in a subtype).

For example, the following test case will start to fail after this change:

```csharp
public interface IFoo
{
	void Method();
}

public class Foo : IFoo
{
	public virtual void Method() { }
}

var mock = new Mock<Foo>();
mock.As<IFoo>().Setup(m => m.Method());  // sets up IFoo.Method
mock.Object.Method();  // calls Foo.Method, which implements IFoo.Method
mock.VerifyAll();  // should therefore pass
```

What your PR has brought into plain sight is the fact that Moq's unit test suite is apparently lacking some very basic tests about method overrides, and how they map to setups. I think that we need to add a rigorous and complete set of unit tests to cover that area before we proceed with fixing this issue.